core/tauri-codegen/src/context.rs

```diff
@@ -289,17 +283,12 @@ pub fn context_codegen(data: ContextData) -> Result<TokenStream, EmbeddedAssetsE
   let with_tray_icon_code = if target.is_desktop() {
     if let Some(tray) = &config.app.tray_icon {
       let tray_icon_icon_path = config_parent.join(&tray.icon_path);
-      let ext = tray_icon_icon_path.extension();
-      if ext.map_or(false, |e| e == "ico") {
-        ico_icon(&root, &out_dir, tray_icon_icon_path)
-          .map(|i| quote!(context.set_tray_icon(Some(#i));))?
-      } else if ext.map_or(false, |e| e == "png") {
-        png_icon(&root, &out_dir, tray_icon_icon_path)
-          .map(|i| quote!(context.set_tray_icon(Some(#i));))?
+      let result = image_icon(&root, &out_dir, &tray_icon_icon_path)
+        .map(|i| quote!(context.set_tray_icon(Some(#i));));
+      if let Err(EmbeddedAssetsError::InvalidImageExtension { .. }) = &result {
+        quote!(compile_error!("The tray icon extension must be either `.ico` or `.png`, but {extension} is given in {path}"))
       } else {
-        quote!(compile_error!(
-          "The tray icon extension must be either `.ico` or `.png`."
-        ))
+        result?
       }
```

Comment:
this could probably be cleaned up into 
```rs
match image_icon() {
  Ok(i) => context.set_tray_icon(),
  Err(EmbeddedAssetsError::InvalidImageExtension {..}) => compile_error!(),
  e => return e, // propagate error
}

```